c2rust-analyze/src/rewrite/expr/unlower.rs

```diff
@@ -0,0 +1,379 @@
+        &'a mir::Operand<'tcx>,
+        &'a [mir::Operand<'tcx>],
+    )> {
+        let &loc = locs.last()?;
+        let term = self.mir.stmt_at(loc).right()?;
+        match term.kind {
+            mir::TerminatorKind::Call {
+                ref func,
+                ref args,
+                destination,
+                ..
+            } => Some((loc, destination, func, args)),
+            _ => None,
+        }
+    }
+
+    fn should_ignore_statement(&self, loc: Location) -> bool {
+        if let Some(stmt) = self.mir.stmt_at(loc).left() {
+            match stmt.kind {
+                mir::StatementKind::FakeRead(..)
+                | mir::StatementKind::StorageLive(..)
+                | mir::StatementKind::StorageDead(..)
+                | mir::StatementKind::Nop => return true,
+                _ => {}
+            }
+        }
+        false
+    }
+
+    fn is_var(&self, pl: mir::Place<'tcx>) -> bool {
+        pl.projection.len() == 0
+    }
+
+    fn visit_expr_inner(&mut self, ex: &'tcx hir::Expr<'tcx>) {
+        let _g = panic_detail::set_current_span(ex.span);
+
+        let mut locs = Vec::with_capacity(1);
+        for &loc in self.span_index.lookup_exact(ex.span) {
+            if !self.should_ignore_statement(loc) {
+                locs.push(loc);
+            }
+        }
+        if locs.len() == 0 {
+            return;
+        }
+
+        let warn = |slf: &Self, desc| {
+            warn!("{}", desc);
+            info!("locs:");
+            for &loc in &locs {
+                slf.mir.stmt_at(loc).either(
+                    |stmt| info!("  {:?}: {:?}", locs, stmt),
+                    |term| info!("  {:?}: {:?}", locs, term),
+                );
+            }
+            info!("span = {:?}", ex.span);
+            info!("expr = {:?}", ex);
+        };
+
+        // Most exprs end with an assignment, storing the result into a temporary.
+        match ex.kind {
+            hir::ExprKind::Assign(pl, rv, _span) => {
+                // For `Assign`, we expect the assignment to be the whole thing.
+                let (loc, _mir_pl, _mir_rv) = match self.get_sole_assign(&locs) {
+                    Some(x) => x,
+                    None => {
+                        warn(self, "expected exactly one StatementKind::Assign");
+                        return;
+                    }
+                };
+                self.record(loc, &[], ex);
+                self.record(loc, &[SubLoc::Dest], pl);
+                self.record(loc, &[SubLoc::Rvalue], rv);
+            }
+
+            hir::ExprKind::Call(_, args) | hir::ExprKind::MethodCall(_, args, _) => {
+                let (loc, _mir_pl, _mir_func, mir_args) = match self.get_last_call(&locs) {
+                    Some((l, pl, f, a)) if self.is_var(pl) => (l, pl, f, a),
+                    _ => {
+                        warn(self, "expected final Call to store into var");
+                        return;
+                    }
+                };
```

Comment:
```suggestion
                let (loc, _mir_pl, _mir_func, mir_args) = match self
                    .get_last_call(&locs)
                    .filter(|(_, pl, _, _)| self.is_var(*pl))
                {
                    Some(x) => x,
                    _ => {
                        warn(self, "expected final Call to store into var");
                        return;
                    }
                };
```

This way we don't need the confusing short variable names like `l`, `f`, `a`.

Reply:
That works well, too!  I forgot about the `@` trick.